src/tpm2_swtpm.c

```diff
@@ -0,0 +1,286 @@
+/* tpm2_swtpm.c
+ *
+ * Copyright (C) 2006-2020 wolfSSL Inc.
+ *
+ * This file is part of wolfTPM.
+ *
+ * wolfTPM is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * wolfTPM is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
+
+
+
+/**
+ * This implements a subset of TPM TCP protocol as described in
+ * "TPM-Rev-2.0-Part-4-Supporting-Routines-01.38-code"
+ *
+ * This is intended for testing with a simulator such as
+ * http://ibmswtpm.sourceforge.net/ or
+ * https://github.com/stefanberger/swtpm
+ */
+
+#ifdef WOLFTPM_SWTPM
+#include <wolftpm/tpm2.h>
+#include <wolftpm/tpm2_swtpm.h>
+#include <wolftpm/tpm2_packet.h>
+
+#include <unistd.h>
+#include <errno.h>
+#include <string.h>
+#include <stdio.h>
+
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netdb.h>
+
+
+#ifndef TPM2_SWTPM_HOST
+#define TPM2_SWTPM_HOST         "localhost"
+#endif
+#ifndef TPM2_SWTPM_PORT
+#define TPM2_SWTPM_PORT         "2321"
+#endif
+
+static TPM_RC SwTpmTransmit(TPM2_CTX* ctx, const void* buffer, ssize_t bufSz)
+{
+    TPM_RC rc = TPM_RC_SUCCESS;
+    ssize_t wrc = 0;
+
+    if (ctx == NULL || ctx->tcpCtx.fd <= 0 || buffer == NULL) {
+        return BAD_FUNC_ARG;
+    }
+
+    wrc = write(ctx->tcpCtx.fd, buffer, bufSz);
+    if (bufSz != wrc) {
+        rc = SOCKET_ERROR_E;
+    }
+
+#ifdef WOLFTPM_DEBUG_VERBOSE
+    if (wrc < 0) {
+        printf("Failed to send the TPM command to fd %d, got errno %d ="
+               "%s\n", ctx->tcpCtx.fd, errno, strerror(errno));
+    }
+#endif
+
+    return rc;
+}
+
+static TPM_RC SwTpmReceive(TPM2_CTX* ctx, void* buffer, size_t rxSz) {
```

Comment:
Coding style... functions braces always on separate line.

Reply:
Done, thought I had caught them all. Thanks for the catch.